Sources/Amplitude/Utilities/HttpClient.swift

```diff
@@ -9,17 +9,19 @@ import Foundation
 
 class HttpClient {
     let configuration: Configuration
-    internal var session: URLSession
+    internal let session: URLSession
 
     init(configuration: Configuration) {
         self.configuration = configuration
-        // shared instance has limitations but think we are not affected
-        // https://developer.apple.com/documentation/foundation/urlsession/1409000-shared
-        self.session = URLSession.shared
+
+        let sessionConfiguration = URLSessionConfiguration.ephemeral
+        sessionConfiguration.httpMaximumConnectionsPerHost = 2
+        self.session = URLSession(configuration: sessionConfiguration)
     }
 
     func upload(events: String, completion: @escaping (_ result: Result<Int, Error>) -> Void) -> URLSessionDataTask? {
         var sessionTask: URLSessionDataTask?
+        let backgroundTaskCompletion = VendorSystem.current.beginBackgroundTask()
```

Comment:
@falconandy is there any downside to running this for all uploads (including the foreground)? I think previously we considering only doing this on background flush.

Reply:
> I think previously we considering only doing this on background flush.

Current API doesn't allow this case - we need to add optional `completion` callback to `Amplitude.flush()` and `EventPlugin.flush()` methods to let use it in `onExitForeground` function.

> is there any downside to running this for all uploads (including the foreground)

I don't know if there are any downside. Possible advantage is that unfinished API calls will be completed when an app goes to background.

Reply:
The `beginBackgroundTask(expirationHandler:)` API is designed to be called at any point in the app's execution. In fact, that's the only way for it to work effectively. From [the docs](https://developer.apple.com/documentation/uikit/uiapplication/1623031-beginbackgroundtask):

> Call this method as early as possible before starting your task, preferably before your app actually enters the background... For example, don’t call this method at the very end of your [applicationDidEnterBackground(_:)](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622997-applicationdidenterbackground) method and expect your app to continue running. If the system is unable to grant the task assertion, it calls your expiration handler.

And:

> You can call this method at any point in your app’s execution. You may also call this method multiple times to mark the beginning of several background tasks that run in parallel. However, each task must be ended separately.